### PR TITLE
Evolve the todo backend to a ReBAC model

### PR DESCRIPTION
**Pull request by oanatmaria**
Fixes https://github.com/aserto-dev/workspace/issues/515
This PR evolves the Todo backend in the following ways:

- Creating a new todo results in creating a resource and a relation
- Deleting a todo results in deleting the resource with its relations
- The Policy is based on policy-rebac (with rebac.check module)
- POST handler uses the Check middleware
- Other handlers (GET, PUT, DELETE) still use the REST middleware
- Resource context matches the Check convention (using object_id)

### CODE REVIEW

File: Controllers/PostTodoController.cs [hunk 1]
```diff
@@ -28,7 +29,8 @@ public PostTodoController(ITodoService todoService, IUserService userService, IM
         }
 
         [HttpPost]
-        [Authorize("Aserto")]
+        [Authorize]
+        [Check("member")]
```
Comment (ronenh) [inline Controllers/PostTodoController.cs]: I would rename "member" to "is-resource-creator". Member doesn't really convey what this check is about.
I was confused about it until I saw the corresponding block in Startup.cs.

I think that `[Check("is-resource-creator")]` would clear things up.

File: Startup.cs [hunk 2]
```diff
@@ -31,6 +39,15 @@ public Startup(IConfiguration configuration)
 
         public IConfiguration Configuration { get; }
 
+        private string CheckPolicyPathMapper(string policyRoot, HttpRequest request)
+        {
+            if (request.Method == "POST")
```
Comment (ronenh) [inline Startup.cs]: What would we do if we had more than one POST endpoint in the service?

File: Startup.cs [hunk 3]
```diff
@@ -58,20 +75,53 @@ public void ConfigureServices(IServiceCollection services)
...
+                options.IdentityMapper = AuthzIdentityContext.Instance.IdentityMapper;
             });
             //end Aserto options handling
 
+            var checkResourceRules = new Dictionary<string, Func<string, HttpRequest, Struct>>();
+
+            checkResourceRules.Add("member", (policyRoot, httpRequest) =>
+            {
+                Struct result = new Struct();
+                if (httpRequest.Method == "POST")
```
Comment (ronenh) [inline Startup.cs]: Why do we need this check? I thought we only use this "member" rule on the POST route?

File: Startup.cs [hunk 4]
```diff
@@ -58,20 +75,53 @@ public void ConfigureServices(IServiceCollection services)
...
+                {
+                    result.Fields.Add("object_id", Value.ForString("resource-creators"));
+                    result.Fields.Add("object_type", Value.ForString("resource-creator"));
+                    result.Fields.Add("relation", Value.ForString("member"));
+                    return result;
+                }
+
+                if (httpRequest.RouteValues.ContainsKey("id"))
```
Comment (ronenh) [inline Startup.cs]: How would the request contain an "id" parameter if the POST route doesn't have any path params?

File: Controllers/PostTodoController.cs [hunk 5]
```diff
@@ -27,8 +28,8 @@ public PostTodoController(ITodoService todoService, IUserService userService, IM
             this.mapper = mapper;
         }
 
-        [HttpPost]
-        [Authorize("Aserto")]
+        [HttpPost]        
+        [Check(objectID: "resource-creators", objectType: "resource-creator", relationName: "member")]
```
Comment (ronenh) [inline Controllers/PostTodoController.cs]: I would rename `relationName` to `relation` for consistency with the arguments of the `Check` API (objectType, ObjectID, relation, subjectType, subjectID).

File: Mapping/AuthzIdentityContext.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            get
+            {
+                return instance;
+            }
+        }
+
+        public Func<ClaimsPrincipal, IEnumerable<string>, IdentityContext> IdentityMapper { get; set; } = CustomIdentityContext;
+
+        private static IdentityContext CustomIdentityContext(ClaimsPrincipal identity, IEnumerable<string> supportedClaimTypes)
```
Comment (ronenh) [inline Mapping/AuthzIdentityContext.cs]: This feels like logic that could be part of the SDK. How would it be different between applications?

File: Services/TodoService.cs [hunk 7]
```diff
@@ -3,21 +3,44 @@
...
-        public TodoService(AppDbContext dbContext)
+        private readonly Aserto.AspNetCore.Middleware.Clients.Directory.V3.Directory directoryClient;
+        private readonly DirectoryConfig opts;
+        public TodoService(IOptions<DirectoryConfig> config, AppDbContext dbContext)
         {
             db = dbContext;
+            this.opts = config.Value;
+            if (!opts.IsValid())
+            {
+                throw new Exception("Invalid config");
```
Comment (ronenh) [inline Services/TodoService.cs]: This isn't a particularly useful error message. Would it make sense to replace `opts.IsValid()` with `opts.Validate()` that throws an exception with a meaningful error message depending on what exactly is invalid?

File: Services/TodoService.cs [hunk 8]
```diff
@@ -3,21 +3,44 @@
...
+            using var loggerFactory = LoggerFactory.Create(builder =>
+            {
+                builder.AddFilter("Microsoft", LogLevel.Warning)
+                .AddFilter("System", LogLevel.Debug)
+                .AddConsole();
+            });
+
+            var options = new AsertoDirectoryOptions(opts.ServiceUrl, opts.APIKey, opts.TenantID, opts.Insecure);
+
+            var optionsInt = Microsoft.Extensions.Options.Options.Create(options);
```
Comment (ronenh) [inline Services/TodoService.cs]: What is this `optionsInt` exactly? Shouldn't the v3.Directory constructor take an `AsertoDirectoryOptions` parameter?

File: Services/UserService.cs [hunk 9]
```diff
@@ -34,26 +34,26 @@ public UserService(IOptions<DirectoryConfig> config)
...
         }
 
         private async Task<GetUserResponse> GetUserBySub(string sub)
         {
             try
             {
-                var getRelationResponse = await directoryClient.GetRelationAsync(subjectType: "user", objType: "identity", objKey: sub, relationName: "identifier", relationObjectType: "identity");
-                if (getRelationResponse.Results.Count == 0)
+                var getRelationResponse = await directoryClient.GetRelationAsync(subjectType: "user", objType: "identity", objId: sub, relationName: "identifier");
+                if (getRelationResponse.Result == null)
```
Comment (ronenh) [inline Services/UserService.cs]: I don't think `Result` can be null. Wouldn't you get back an error if the relation isn't found?

File: Controllers/DeleteTodoController.cs [hunk 10]
```diff
@@ -24,7 +25,7 @@ public DeleteTodoController(ITodoService todoService, IMapper mapper)
     }
 
     [HttpDelete]
-    [Authorize("Aserto")]
+    [Aserto]
```
Comment (ronenh) [inline Controllers/DeleteTodoController.cs:28]: Is it possible to use this with a named configuration? If an application needs to use two asserto middleware (with different policies, for example) could it do that (e.g. `[Aserto("config1")]`)?

File: Services/UserService.cs [hunk 11]
```diff
@@ -5,23 +5,23 @@
...
+        private readonly Aserto.AspNetCore.Middleware.Clients.Directory.V3.Directory directoryClient;
         private readonly DirectoryConfig opts;
 
         public UserService(IOptions<DirectoryConfig> config)
         {
             this.opts = config.Value;
             if (!opts.IsValid())
             {
-                throw new Exception("Invalid config");
+                throw new Exception("Invalid config. Service url can not be empty");
```
Comment (ronenh) [inline Services/UserService.cs:24]: How do we know that the service URL is empty? Couldn't there be some other reason why `IsValid()` returns false?

File: Aserto.TodoApp.csproj [hunk 12]
```diff
@@ -5,14 +5,18 @@
...
     <PackageReference Include="AutoMapper" Version="11.0.1" />
     <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="11.0.0" />
     <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="3.1.19" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="6.0.6" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.6" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="7.0.15" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="../aserto-dotnet/src/Aserto.AspNetCore.Middleware/Aserto.AspNetCore.Middleware.csproj" />
```
Comment (ronenh) [inline Aserto.TodoApp.csproj]: This needs to be removed.

File: Services/UserService.cs (hunk 11 above)
Comment (oanatmaria) [reply Services/UserService.cs:24]: Is the only thing that we check

File: Aserto.TodoApp.csproj (hunk 12 above)
Comment (oanatmaria) [reply Aserto.TodoApp.csproj]: Only after the sdk is released

**Review by ronenh — APPROVED** [review]

**Pull request merged**